runtime/server/psql.go

```diff
@@ -0,0 +1,252 @@
+// The metrics_sql queries are redirected to metrics_sql resolver.
+func (s *Server) psqlQueryHandler(ctx context.Context, query string) (wire.PreparedStatements, error) {
+	clientParams := wire.ClientParameters(ctx)
+	instanceID := clientParams[wire.ParamDatabase]
+
+	// The logic to identify metadata queries is somewhat hacky.
+	// Identify metadata queries based on whether query contains names of common postgres catalogs like pg_attribute, pg_catalog etc.
+	// Also consider queries that do not have a FROM clause as metadata queries since querying from metrics_view requires a FROM clause.
+	if strings.Contains(query, "pg_attribute") || strings.Contains(query, "pg_catalog") || strings.Contains(query, "pg_type") || strings.Contains(query, "pg_namespace") || !strings.Contains(strings.ToUpper(query), "FROM") {
+		data, schema, err := psql.ResolvePSQLQuery(ctx, &psql.PSQLQueryOpts{
+			SQL:            query,
+			Runtime:        s.runtime,
+			InstanceID:     instanceID,
+			UserAttributes: auth.GetClaims(ctx).Attributes(),
+			Priority:       1,
+			Logger:         s.logger,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		handle := func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
+			for i := 0; i < len(data); i++ {
+				if err := writer.Row(data[i]); err != nil {
+					return fmt.Errorf("failed to write row: %w", err)
+				}
+			}
+			return writer.Complete("OK")
+		}
+		return wire.Prepared(wire.NewStatement(handle, wire.WithColumns(convertSchema(schema)))), nil
+	}
+
+	// todo handle normal SQL. Some ideas to handle that
+	// 1. Approach similar to BigQuery which has #legacySQL
+	// 2. We can parse for a -- @connector: xxx comment, and if found, use that connector directly instead of metrics SQL
+	res, err := s.runtime.Resolve(ctx, &runtime.ResolveOptions{
+		InstanceID:         instanceID,
+		Resolver:           "metrics_sql",
+		ResolverProperties: map[string]any{"sql": query},
+		Args:               map[string]any{"priority": 1},
+		UserAttributes:     auth.GetClaims(ctx).Attributes(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return wire.Prepared(wire.NewStatement(s.execQuery(instanceID, query), wire.WithColumns(convertSchema(res.Schema)))), nil
+}
+
+func (s *Server) execQuery(instanceID, query string) wire.PreparedStatementFn {
+	return func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
+		// NOTE :: the duplicate call to runtime.Resolve is made so that we do not keep the results and eventually leak memory if query is never executed.
+		// This should resolve fast in most cases since results will be cached.
+		res, err := s.runtime.Resolve(ctx, &runtime.ResolveOptions{
+			InstanceID:         instanceID,
+			Resolver:           "metrics_sql",
+			ResolverProperties: map[string]any{"sql": query},
+			Args:               map[string]any{"priority": 1},
+			UserAttributes:     auth.GetClaims(ctx).Attributes(),
+		})
```

Comment:
We only cache for DuckDB, not Druid or ClickHouse. I think running each psql query twice, even if it may lead into DB internal caches is too excessive. If using `s.runtime.Resolve`, which returns JSON, why can't that just be passed in here? There is risk of a leaked connection then as far as I can tell?